packages/react-instantsearch-hooks-dom/src/ui/RangeInput.tsx

```diff
@@ -0,0 +1,174 @@
+import React from 'react';
+
+import { cx } from './lib/cx';
+
+import type { useRange } from 'react-instantsearch-hooks';
+
+type RangeRenderState = ReturnType<typeof useRange>;
+
+export type RangeInputProps = Omit<
+  React.HTMLAttributes<HTMLDivElement>,
+  'onSubmit'
+> &
+  Pick<RangeRenderState, 'range' | 'start'> & {
+    classNames?: Partial<RangeInputClassNames>;
+    disabled: boolean;
+    onSubmit: RangeRenderState['refine'];
+    step?: number;
+    translations: RangeInputTranslations;
+  };
+
+export type RangeInputClassNames = {
+  /**
+   * Class names to apply to the root element
+   */
+  root: string;
+  /**
+   * Class names to apply to the root element when there are no refinements possible
+   */
+  rootNoRefinement: string;
+  /**
+   * Class names to apply to the form element
+   */
+  form: string;
+  /**
+   * Class name to apply to each label element
+   */
+  label: string;
+  /**
+   * Class names to apply to each input element
+   */
+  input: string;
+  /**
+   * Class names to apply to the minimum input element
+   */
+  inputMin: string;
+  /**
+   * Class names to apply to the maximum input element
+   */
+  inputMax: string;
+  /**
+   * Class names to apply to the separator element
+   */
+  separator: string;
+  /**
+   * Class names to apply to the submit button
+   */
+  submit: string;
+};
+
+export type RangeInputTranslations = {
+  /**
+   * The label of the separator, between the minimum and maximum inputs
+   */
+  separator: string;
+  /**
+   * The label of the submit button
+   */
+  submit: string;
+};
+
+// if the default value is undefined, React considers the component uncontrolled initially, which we don't want 0 or NaN as the default value
+const unsetNumberInputValue = '' as unknown as number;
+
+export function RangeInput({
+  classNames = {},
+  range: { min, max },
+  start: [minValue, maxValue],
+  step = 1,
+  disabled,
+  onSubmit,
+  translations,
+  ...props
+}: RangeInputProps) {
+  const values = {
+    min:
+      minValue !== -Infinity && minValue !== min
+        ? minValue
+        : unsetNumberInputValue,
+    max:
+      maxValue !== Infinity && maxValue !== max
+        ? maxValue
+        : unsetNumberInputValue,
+  };
+
+  const [{ from, to }, setRange] = React.useState({
+    from: values.min,
+    to: values.max,
+  });
+
+  React.useEffect(() => {
+    setRange({ from: values.min, to: values.max });
+  }, [values.min, values.max]);
+
+  const onInput =
+    (key: string) =>
+    ({ currentTarget }: React.SyntheticEvent<HTMLInputElement>) => {
+      const value = Number(currentTarget.value);
+
+      currentTarget.value = value.toString();
```

Comment:
What's this line for? Tests don't fail without it.

Reply:
This line is what's allowing the overwrite of the `0` when it prefixes other digits (see [sandbox](https://codesandbox.io/s/recursing-golick-9hn904?file=/src/App.js)).

![CleanShot 2022-04-11 at 11 59 42](https://user-images.githubusercontent.com/154633/162715257-6892d261-a699-4817-8dcd-9c26da6cd592.gif)

I have issue with finding a good way to test that as the `.toHaveValue()` assertion expects a number, and directly checking the `value` of the input returns the expected value, whether or not the line is present.

Do you have any ideas as to why?

Reply:
Instead of mutating the target value, we can rely on the `value` HTML attribute to stringify:

```tsx
<input
  // ...
  value={from?.toString()}
/>
```

As for the test, it's indeed quite a browser behavior so I'm not sure without digging into internals. I would suggest adding a comment as an alternative.

Reply:
Thanks, that works! I'm guessing its due to the fact that now React checks the value as a string instead of a number.